### PR TITLE
[WIP] Add Carla Env

### PR DESCRIPTION
**Pull request by RPC2**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -51,3 +51,4 @@ git+git://github.com/HumanCompatibleAI/stable-baselines3@ad902dd5f9d4afeef347897
 magical-il~=0.0.1a4
 dm_control~=0.0.319497192
 git+git://github.com/denisyarats/dmc2gym@6e34d8acf18e92f0ea0a38ecee9564bdf2549076#egg=dmc2gym
+git+https://github.com/rail-berkeley/d4rl@master#egg=d4rl
```
Comment (qxcv) [inline requirements.txt:54]: I think this also needs `pygame`, `networkx` and `dotmap` (for CARLA specifically).

File: src/il_representations/algos/encoders.py [hunk 2]
```diff
@@ -156,6 +157,39 @@ def do_warning(message):
...
+    """Use MLP for encodder network, with structure defined by model_arch.
+       Its default follows https://github.com/rail-berkeley/d4rl_evaluations/b
+       lob/c93b9d2bf443cd5b8c43542fca9937dc50e630c3/brac/behavior_regularized_
+       offline_rl/brac/networks.py#L39.
+    """
+    def __init__(self, observation_space, representation_dim, model_arch,
+                 flatten_input=False):
+        super().__init__()
+        if flatten_input:
+            # Assuming the input has frame stack. We only use the last RGB frame.
```
Comment (qxcv) [inline src/il_representations/algos/encoders.py:170]: This is a bit dangerous—it's silently throwing away frame stacking, which might cause this architecture to mysteriously fail in partially observable domains.
Why not flatten directly, without this behaviour? It won't affect CARLA at all (since we're not stacking frames), and is less likely to cause confusing results in other domains.

File: src/il_representations/algos/encoders.py [hunk 3]
```diff
@@ -156,6 +157,39 @@ def do_warning(message):
...
+        for count, dim in enumerate(model_arch):
+            out_dim = dim
+            layer_dict[f'fc{count}'] = nn.Linear(in_dim, out_dim)
+            layer_dict[f'relu{count}'] = nn.ReLU()
+            in_dim = out_dim
+        self.shared_network = nn.Sequential(layer_dict)
+
+    def forward(self, x):
+        if self.flatten_input:
+            x = x[:, -3:, ...].flatten(start_dim=1)
```
Comment (qxcv) [inline src/il_representations/algos/encoders.py:191]: Similar comment about flattening.

File: src/il_representations/envs/auto.py [hunk 4]
```diff
@@ -195,6 +202,15 @@ def load_vec_env(benchmark_name, dm_control_full_env_names,
         assert final_env.observation_space.shape == (12, 64, 64), \
             final_env.observation_space.shape
         return final_env
+    elif benchmark_name == 'carla':
+        raw_carla_env = gym.make(gym_env_name)
+        raw_carla_env = CarlaImageObsEnv(raw_carla_env)
+        raw_carla_env = VecMonitor(venv=raw_carla_env, filename=None,
+                                            keep_buf=100)
+        final_env = VecFrameStack(raw_carla_env, 4, 'first')
```
Comment (qxcv) [inline src/il_representations/envs/auto.py:210]: Is frame stacking used in the D4RL paper? I thought their default datasets only had 48*48*3 observations, not 48*4*12.

File: src/il_representations/envs/auto.py [hunk 5]
```diff
@@ -208,7 +224,7 @@ def get_data_dir(benchmark_name, task_key, data_type, data_root):
     """Get the data directory for a given benchmark ("magical", "dm_control",
     etc.), task (e.g. "MoveToCorner-Demo-v0", "finger-spin") and data type
     (e.g. "demos", "random")."""
-    return os.path.join(data_root, 'data', 'processed',
+    return os.path.join(data_root, 'carla', 'data', 'processed',
```
Comment (qxcv) [inline src/il_representations/envs/auto.py:227]: Why is this needed?

File: src/il_representations/envs/auto.py (hunk 4 above)
Comment (qxcv) [reply src/il_representations/envs/auto.py:210]: Also, I noticed that `load_dataset_carla` takes a `carla_frame_stack` config var. If that's the case, then this function should use `carla_frame_stack` instead of a constant of 4.

File: src/il_representations/envs/carla_envs.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    trajectories = np.load(full_rollouts_path, allow_pickle=True)
+
+    obs = trajectories['observations'].astype(np.uint8)
+    nobs = trajectories['next_observations'].astype(np.uint8)
+    acts = trajectories['actions'].astype(np.uint8)
+    rews = trajectories['rewards'].astype(np.uint8)
+    dones = trajectories['terminals'].astype(np.uint8)
+
+    if n_traj is not None:
+        nth_traj_end_idx = [i for i, n in enumerate(cat_dones) if n][n_traj-1] + 1
```
Comment (qxcv) [inline src/il_representations/envs/carla_envs.py:33]: Can also use `np.flatnonzero(cat_dones)[n_traj-1]+1`.

File: src/il_representations/envs/carla_envs.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+
+@env_cfg_ingredient.capture
+def load_dataset_carla(task_name, carla_frame_stack, chans_first=True,
+                       n_traj=None):
+    data_root, carla_demo_paths = _get_carla_data_opts()
+
+    # load trajectories from disk
+    full_rollouts_path = os.path.join(data_root, carla_demo_paths[task_name])
+    trajectories = np.load(full_rollouts_path, allow_pickle=True)
```
Comment (qxcv) [inline src/il_representations/envs/carla_envs.py:24]: I take it that this is not the native D4RL data format? If so, how were these trajectories saved/generated?

File: src/il_representations/envs/carla_envs.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    def reset(self):
+        obs = self.wrapped_env.reset()
+        # This is BCHW
+        obs = obs.reshape([1] + list(self.observation_space.shape))
+        obs = (obs * 255).astype(np.uint8)
+        return obs
+
+    def step(self, action):
+        next_obs, reward, done, info = self.wrapped_env.step(action)
+        next_obs = next_obs.reshape([1] + list(self.observation_space.shape))
```
Comment (qxcv) [inline src/il_representations/envs/carla_envs.py:91]: Why is the leading [1] necessary?

File: src/il_representations/envs/carla_envs.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    dataset_dict['obs'] = stack_obs_oldest_first(dataset_dict['obs'],
+                                                 carla_frame_stack,
+                                                 use_zeroed_frames=True)
+    dataset_dict['next_obs'] = stack_obs_oldest_first(dataset_dict['next_obs'],
+                                                 carla_frame_stack,
+                                                 use_zeroed_frames=True)
+
+    return dataset_dict
+
+class CarlaImageObsEnv:
```
Comment (qxcv) [inline src/il_representations/envs/carla_envs.py:61]: This should inherit from `gym.Wrapper` so that it gets all the standard env methods. Probably `gym.ObservationWrapper` is the ideal parent class from the perspective of minimising code required (you only have to override `__init__` to set the `observation_space`, then override `observation(x)` method to reshape observations).

**Review by qxcv — COMMENTED** [review]
Some comments, but I'm pretty sure none of them should affect our CARLA results, so they can wait until after the rebuttal deadilne.

File: requirements.txt (hunk 1 above)
Comment (RPC2) [reply requirements.txt:54]: Added!
